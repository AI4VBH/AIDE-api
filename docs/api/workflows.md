# Workflows

Manage workflows in the Workflow Manager.

## `GET /workflows?pageNumber=:pageNumber&pageSize=:pageSize`

Returns a paginated list of workflows.

### Parameters

| Parameter Type | Label | Description |
|----------------|-------|-------------|
| Query | `:pageNumber` | `number` current page number. default `0` |
| Query | `:pageSize` | `number` the number of items per page. default `10` |

### Responses

| Code | Description |
|------|-------------|
| `200` | Returns the paginated list of workflows. `PagedMonaiWorkflows` |
| `400` | Invalid range for `pageNumber` or `pageSize` |
| `500` | Server error. The response will be a [Problem details](https://datatracker.ietf.org/doc/html/rfc7807) object with server error details. |

### Example Request

```bash
curl --location --request GET 'http://localhost:5000/workflows?pageNumber=1&pageSize=5'
```

### Example Response

```json
{
  "totalPages": 1,
  "totalRecords": 1,
  "data": [
    {
      "id": "1564c53c-53b7-4971-8bb2-a3d3bc051f1c",
      "workflow_id": "5507c486-5a28-4f93-a0a3-0014c308318f",
      "revision": 1,
      "name": "workflow-name",
      "description": "workflow description",
      "version": "v1.0"
    }
  ]
}
```

---

## `GET /workflows/:workflowId`

Returns the workflow information from MONAI.

### Parameters

| Parameter Type | Label | Description |
|----------------|-------|-------------|
| URL | `:workflowId` | the id of the workflow |

### Responses

| Code | Description |
|------|-------------|
| `200` | The workflow object from MONAI. |
| `404` | Workflow with the given ID not found. |
| `500` | Server error. The response will be a [Problem details](https://datatracker.ietf.org/doc/html/rfc7807) object with server error details. |

### Example Request

```bash
curl --location --request GET 'http://localhost:5000/workflows/6542f955-8ddb-4e91-9ca3-948b1e32ffd9'
```

### Example Response

```json
{
  "id": "f58fe206-1976-45bc-9d14-b9459eee2dfe",
  "revision": 3,
  "workflow": {
    "description": "This workflow is to be used for connectathon",
    "informatics_gateway": {
      "ae_title": "MonaiSCU",
      "data_origins": [
        "test",
      ],
      "export_destinations": [
        "ORTHANC",
      ],
    },
    "name": "Export Workflow",
    "tasks": [
      {
        "args": {
          "test": "test1",
        },
        "artifacts": {
          "input": [
            {
              "mandatory": true,
              "name": "study",
              "value": "{{ context.input.dicom }}",
            },
          ],
          "output": [],
        },
        "description": "Connectathon export task",
        "export_destinations": [
          {
            "name": "ORTHANC",
          },
        ],
        "id": "export-task-connectathon",
        "input_parameters": null,
        "ref": "",
        "task_destinations": [],
        "type": "export",
      },
    ],
    "version": "2.0.0",
  },
  "workflow_id": "0ea7b5b9-64ba-4841-b252-d6e312ef7e8e",
}
```

---

## `POST /workflows`

Create a new workflow in the Workflow Manager.

### Parameters

| Parameter Type | Label | Description |
|----------------|-------|-------------|
| Body | | `CreateEditWorkflowDto`. workflow that needs creating |

### Responses

| Code | Description |
|------|-------------|
| `201` | Workflow created. |
| `400` | Invalid workflow object. |
| `500` | Server error. The response will be a [Problem details](https://datatracker.ietf.org/doc/html/rfc7807) object with server error details. |

### Example Request

```bash
# rest of workflow body removed for brevity

curl --location --request POST 'http://localhost:5000/workflows' \
  --header 'Content-Type: application/json' \
  --data-raw '{
    "workflow": {
      "informatics_gateway": {
        "ae_title": "TITLE",
        "export_destinations": []
      }
    }
  }'
```

### Example Response



---

## `PUT /workflows/:workflowId`

Create a new workflow in the Workflow Manager.

### Parameters

| Parameter Type | Label | Description |
|----------------|-------|-------------|
| URL | `:workflowId` | Id of workflow to edit |
| Body | | `CreateEditWorkflowDto`. workflow object |

### Responses

| Code | Description |
|------|-------------|
| `201` | Workflow updated. |
| `400` | Invalid workflow object. |
| `404` | Workflow with the given ID not found |
| `500` | Server error. The response will be a [Problem details](https://datatracker.ietf.org/doc/html/rfc7807) object with server error details. |

### Example Request

```bash
# rest of workflow body removed for brevity

curl --location --request PUT 'http://localhost:5000/workflows/workflow-id' \
  --header 'Content-Type: application/json' \
  --data-raw '{
    "workflow": {
      "informatics_gateway": {
        "ae_title": "TITLE",
        "export_destinations": []
      }
    }
  }'
```

---

## `DELETE /workflows/:workflowId`

Delete the workflow with a given ID.

### Parameters

| Parameter Type | Label | Description |
|----------------|-------|-------------|
| URL | `:workflowId` | Id of workflow to delete |

### Responses

| Code | Description |
|------|-------------|
| `200` | Workflow deleted. |
| `404` | Workflow with the given ID not found |
| `500` | Server error. The response will be a [Problem details](https://datatracker.ietf.org/doc/html/rfc7807) object with server error details. |

### Example Request

```bash
curl --location --request DELETE 'http://localhost:5000/workflows/workflow-id'
```

### Example Response

N/A