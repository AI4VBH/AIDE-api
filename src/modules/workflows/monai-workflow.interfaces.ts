import { IPagedResponse } from 'shared/helper/paging/paging.interface';

export type PagedMonaiWorkflows = IPagedResponse<MonaiWorkflow>;

export interface MonaiWorkflow {
  id: string;
  workflow_id: string;
  revision: number;
  workflow: {
    name: string;
    version: string;
    description: string;
  };
}
