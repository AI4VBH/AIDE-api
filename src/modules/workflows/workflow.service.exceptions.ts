export class WorkflowServiceException extends Error {}
