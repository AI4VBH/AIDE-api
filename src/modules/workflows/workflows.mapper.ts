import {
  PagedWorkflowsDto,
  PagedWorkflowsItemDto,
} from './dto/aide-workflow.dto';
import {
  MonaiWorkflow,
  PagedMonaiWorkflows,
} from './monai-workflow.interfaces';

export function mapToPagedWorkflowsDto(
  paged: PagedMonaiWorkflows,
): PagedWorkflowsDto {
  if (!paged) {
    throw new Error('paged is null or undefined');
  }

  return {
    totalPages: paged.totalPages,
    totalRecords: paged.totalRecords,
    data: paged.data.map((w) => mapToPagedWorkflowsItemDto(w)),
  };
}

export function mapToPagedWorkflowsItemDto(
  workflow: MonaiWorkflow,
): PagedWorkflowsItemDto {
  return {
    id: workflow.id,
    workflow_id: workflow.workflow_id,
    revision: workflow.revision,
    name: workflow.workflow.name,
    description: workflow.workflow.description,
    version: workflow.workflow.version,
    ae_title: workflow.workflow.informatics_gateway.ae_title,
    data_origins: workflow.workflow.informatics_gateway.data_origins,
  };
}
