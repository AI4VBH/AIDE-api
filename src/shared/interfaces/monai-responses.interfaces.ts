export interface IMonaiProblemResponse {
  type: string;
  title: string;
  status: number;
  detail: string;
  instance: string;
  traceId: string;
}
