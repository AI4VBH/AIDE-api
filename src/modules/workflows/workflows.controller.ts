import {
  BadRequestException,
  Body,
  Controller,
  Delete,
  Get,
  Inject,
  Param,
  ParseIntPipe,
  Post,
  Put,
  Query,
  UseFilters,
} from '@nestjs/common';
import MonaiServerExceptionFilter from 'shared/http/monai-server-exception.filter';
import { CreateEditWorkflowDto } from './dto/aide-workflow.dto';
import { WorkflowsService } from './workflows.service';

@Controller('workflows')
@UseFilters(MonaiServerExceptionFilter)
export class WorkflowsController {
  @Inject(WorkflowsService)
  private readonly service: WorkflowsService;

  @Get()
  getWorkflows(
    @Query('pageNumber', ParseIntPipe) pageNumber = 1,
    @Query('pageSize', ParseIntPipe) pageSize = 10,
  ) {
    if (pageNumber <= 0 || pageSize <= 0) {
      throw new BadRequestException('pageNumber or pageSize is invalid');
    }

    return this.service.getPagedWorkflows(pageNumber, pageSize);
  }

  @Get(':workflowId')
  getWorkflowById(@Param('workflowId') workflowId: string) {
    return this.service.getWorkflowDetail(workflowId);
  }

  @Post()
  createWorkflow(@Body() createWorkflow: CreateEditWorkflowDto) {
    return this.service.createWorkflow(createWorkflow.workflow);
  }

  @Put(':workflowId')
  editWorkflow(
    @Param('workflowId') workflowId: string,
    @Body() editWorkflow: CreateEditWorkflowDto,
  ) {
    return this.service.editWorkflow(workflowId, editWorkflow.workflow);
  }

  @Delete(':workflowId')
  deleteWorkflowById(@Param('workflowId') workflowId: string) {
    return this.service.deleteWorkflow(workflowId);
  }
}
