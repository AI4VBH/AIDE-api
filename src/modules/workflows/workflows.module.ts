import { HttpModule } from '@nestjs/axios';
import { Module } from '@nestjs/common';
import { HttpConfigService } from 'shared/http/http.service';
import { WorkflowsController } from './workflows.controller';
import { WorkflowsService } from './workflows.service';

@Module({
  imports: [
    HttpModule.registerAsync({
      useClass: HttpConfigService,
    }),
  ],
  controllers: [WorkflowsController],
  providers: [WorkflowsService],
})
export class WorkflowsModule {}
