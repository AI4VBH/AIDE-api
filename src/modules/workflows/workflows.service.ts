import { HttpService } from '@nestjs/axios';
import { HttpStatus, Inject, Injectable } from '@nestjs/common';
import { firstValueFrom } from 'rxjs';
import {
  MonaiWorkflow,
  PagedMonaiWorkflows,
} from './monai-workflow.interfaces';
import { mapToPagedWorkflowsDto } from './workflows.mapper';

@Injectable()
export class WorkflowsService {
  @Inject(HttpService)
  private readonly httpService: HttpService;

  async getPagedWorkflows(pageNumber: number, pageSize: number) {
    const params = new URLSearchParams({
      pageSize: `${pageSize || ''}`,
      pageNumber: `${pageNumber || ''}`,
    });

    const paged = await firstValueFrom(
      this.httpService.get<PagedMonaiWorkflows>(`/workflows?${params}`),
    );

    return mapToPagedWorkflowsDto(paged.data);
  }

  async getWorkflowDetail(workflowId: string) {
    const detail = await firstValueFrom(
      this.httpService.get<MonaiWorkflow>(`/workflows/${workflowId}`),
    );

    return detail.data;
  }

  async createWorkflow(workflow: object) {
    const result = await firstValueFrom(
      this.httpService.post('/workflows', workflow),
    );

    return result.data;
  }

  async editWorkflow(workflowId: string, workflow: object) {
    const result = await firstValueFrom(
      this.httpService.put(`/workflows/${workflowId}`, workflow),
    );

    return result.data;
  }

  async deleteWorkflow(workflowId: string) {
    const result = await firstValueFrom(
      this.httpService.delete(`/workflows/${workflowId}`),
    );

    return result.data;
  }
}
