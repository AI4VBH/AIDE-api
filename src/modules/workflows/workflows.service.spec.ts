import { createMock, DeepMocked } from '@golevelup/ts-jest';
import { HttpService } from '@nestjs/axios';
import { Test, TestingModule } from '@nestjs/testing';
import { AxiosInstance } from 'axios';
import { makeObservableForTest } from '../../../test/utilities/test-make-observable';
import { PagedMonaiWorkflows } from './monai-workflow.interfaces';
import { WorkflowsService } from './workflows.service';

describe('WorkflowsService', () => {
  let service: WorkflowsService;
  let axios: DeepMocked<AxiosInstance>;
  let httpService: DeepMocked<HttpService>;

  beforeEach(async () => {
    axios = createMock<AxiosInstance>();
    httpService = createMock<HttpService>({
      axiosRef: axios,
    });

    const module: TestingModule = await Test.createTestingModule({
      providers: [
        WorkflowsService,
        {
          provide: HttpService,
          useValue: httpService,
        },
      ],
    }).compile();

    service = module.get<WorkflowsService>(WorkflowsService);
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  describe('getPagedWorkflows', () => {
    it('returns expected result', async () => {
      const monaiResult: PagedMonaiWorkflows = {
        totalPages: 1,
        totalRecords: 1,
        pageNumber: 1,
        pageSize: 10,
        firstPage: '/page-1',
        lastPage: '/page-1',
        nextPage: null,
        previousPage: null,
        data: [],
      };

      axios.get.mockResolvedValue({
        status: 200,
        data: monaiResult,
      });

      httpService.get.mockReturnValue(makeObservableForTest(axios.get));

      const result = await service.getPagedWorkflows(1, 10);

      expect(result).toMatchSnapshot();
    });
  });
});
