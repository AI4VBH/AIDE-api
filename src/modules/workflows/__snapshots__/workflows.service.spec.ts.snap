// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WorkflowsService getPagedWorkflows returns expected result 1`] = `
{
  "data": [],
  "totalPages": 1,
  "totalRecords": 1,
}
`;
